ui/src/criteria/attribute.tsx

```diff
@@ -81,23 +133,145 @@ function AttributeEdit(props: AttributeEditProps) {
 
-  const selectionIndex = (hint: tanagra.EnumHintValue) =>
-    props.data.selected.findIndex(
-      (row) => row.id === hint.attributeValue?.int64Val
-    );
+  const hasValidIntegerBounds =
+    isValid(integerBoundsHint?.min) && isValid(integerBoundsHint?.max);
 
-  if (enumHintValues?.length === 0) {
+  if (!hasValidIntegerBounds && enumHintValues?.length === 0) {
     return (
       <Typography>
         No information for attribute {props.data.attribute}.
       </Typography>
     );
   }
+
+  if (hasValidIntegerBounds) {
+    const minBound = integerBoundsHint?.min || 0; // This is to ensure the compiler won't complain the object might be undefined
+    const maxBound = integerBoundsHint?.max || 0; // although we already know that min and max is valid
+
+    //two sets of values are needed due to the input box and slider is isolated.
+    const [minInputValue, setMinInputValue] = useState<number>(minBound);
+    const [maxInputValue, setMaxInputValue] = useState<number>(maxBound);
+    const [minValue, setMinValue] = useState<number>(minBound);
+    const [maxValue, setMaxValue] = useState<number>(maxBound);
+
+    const updateValue = (newMin: number, newMax: number) => {
+      setMinValue(newMin);
+      setMaxValue(newMax);
+      props.dispatchFn(
+        produce(props.data, (data) => {
+          data.min = newMin;
+          data.max = newMax;
+        })
+      );
+    };
+
+    const handleChange = (event: Event, newValue: number | number[]) => {
+      newValue = newValue as number[];
+      const [newMin, newMax] = newValue;
+      setMinInputValue(newMin);
+      setMaxInputValue(newMax);
+      updateValue(newMin, newMax);
+    };
+
+    const handleLeftInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMin =
+        event.target.value === "" ? minBound : Number(event.target.value);
+      setMinInputValue(newMin);
+
+      newMin = Math.max(minBound, newMin);
+      newMin = Math.min(maxValue, newMin);
+      updateValue(newMin, maxValue);
+    };
+    const handleLeftInputBlur = () => {
+      let newMin = minInputValue;
+      newMin = Math.max(newMin, minBound);
+      newMin = Math.min(newMin, maxValue);
+      setMinInputValue(newMin);
+    };
+
+    const handleRightInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMax =
+        event.target.value === "" ? maxBound : Number(event.target.value);
+      setMaxInputValue(newMax);
+
+      newMax = Math.max(minValue, newMax);
+      newMax = Math.min(newMax, maxBound);
+      updateValue(minValue, newMax);
+    };
+    const handleRightInputBlur = () => {
+      let newMax = maxInputValue;
+      newMax = Math.max(minValue, newMax);
+      newMax = Math.min(newMax, maxBound);
+      setMaxInputValue(newMax);
+    };
+
+    return (
+      <Box sx={{ width: "30%", minWidth: 100 }}>
+        <Grid container spacing={2} alignItems="center">
```

Comment:
Yes. I find It should be L221 add a margin. then it will be good. I set the margin as 2. and looks better
<img width="707" alt="Screen Shot 2022-04-20 at 5 09 38 PM" src="https://user-images.githubusercontent.com/100877272/164323465-04e1ce86-b723-4f9f-abfe-c4bcfdc0b13f.png">